lib/propolis/src/tasks.rs

```diff
@@ -0,0 +1,373 @@
+        guard.wake_ctrl();
+        self.cv.notify_all();
+    }
+    fn set_unheld(&self, guard: &mut MutexGuard<Control>) {
+        guard.is_held = false;
+        guard.waker_task = None;
+        // Notify ctrl of changed state
+        guard.wake_ctrl();
+        self.cv.notify_all();
+    }
+}
+
+#[derive(Copy, Clone, Debug)]
+pub enum Event {
+    Hold,
+    Exit,
+}
+
+pub struct TaskHdl(Arc<Inner>);
+impl TaskHdl {
+    pub fn new(notify_fn: Option<Box<NotifyFn>>) -> (Self, TaskCtrl) {
+        Self::new_inner(false, notify_fn)
+    }
+    pub fn new_held(notify_fn: Option<Box<NotifyFn>>) -> (Self, TaskCtrl) {
+        Self::new_inner(true, notify_fn)
+    }
+    fn new_inner(
+        initial_hold: bool,
+        notify_fn: Option<Box<NotifyFn>>,
+    ) -> (Self, TaskCtrl) {
+        let inner = Arc::new(Inner {
+            ctrl: Mutex::new(Control {
+                should_hold: initial_hold,
+                ..Default::default()
+            }),
+            cv: Condvar::new(),
+            notify_fn,
+        });
+        let ctrl = TaskCtrl(Arc::downgrade(&inner));
+        (Self(inner), ctrl)
+    }
+
+    pub fn pending_event(&self) -> Option<Event> {
+        let ctrl = self.0.ctrl.lock().unwrap();
+        if ctrl.should_exit {
+            Some(Event::Exit)
+        } else if ctrl.should_hold {
+            Some(Event::Hold)
+        } else {
+            None
+        }
+    }
+    pub fn hold(&self) {
+        let mut guard = self.0.ctrl.lock().unwrap();
+        if guard.should_exit || !guard.should_hold {
+            return;
+        }
+        assert!(!guard.is_held, "task already held");
+        guard.is_held = true;
+        let cv = &self.0.cv;
+        cv.notify_all();
+        let mut guard =
+            cv.wait_while(guard, |g| g.should_hold && !g.should_exit).unwrap();
+        guard.is_held = false;
+    }
+    pub fn force_hold(&self) {
+        let mut guard = self.0.ctrl.lock().unwrap();
+        if guard.should_exit {
+            return;
+        }
+        // induce ourself into the held state
+        guard.should_hold = true;
+        assert!(!guard.is_held, "task already held");
+        guard.is_held = true;
+        let cv = &self.0.cv;
+        cv.notify_all();
+        let mut guard =
+            cv.wait_while(guard, |g| g.should_hold && !g.should_exit).unwrap();
+        guard.is_held = false;
+    }
+
+    pub async fn wait_held(&mut self) {
+        Held::new(self).await
+    }
+    pub async fn get_event(&mut self) -> Event {
+        GetEvent::new(self).await
+    }
+}
+
+struct Held<'a> {
+    hdl: &'a mut TaskHdl,
+    held: bool,
+}
+impl<'a> Held<'a> {
+    fn new(hdl: &'a mut TaskHdl) -> Self {
+        Self { hdl, held: false }
+    }
+}
+impl Future for Held<'_> {
+    type Output = ();
+
+    fn poll(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Self::Output> {
+        let mut guard = self.hdl.0.ctrl.lock().unwrap();
```

Comment:
We should not be blocking in any `poll` method:

From the [std docs](https://doc.rust-lang.org/stable/std/future/trait.Future.html#runtime-characteristics):
> An implementation of poll should strive to return quickly, and should not block.

In practise this might be fine but deadlocked tasks are a pain to debug.

Reply:
I'm not especially concerned about this.  No significantly lengthy work is being done when the lock is being held, but using a traditional Mutex makes implementation of the synchronous side of this interface considerably easier.  I'm certainly not against heeding the guidance in `std`, but it is SHOULD rather than MUST here.  There are certainly other ways that async consumers may stumble over blocking operations (really anything touching the kernel vmm may block on in-kernel mutex during an ioctl)

Reply:
I've added a comment to the mutex highlighting the care which must be taken